configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionRepository.java

```diff
@@ -682,6 +682,77 @@ private boolean sessionLifeTimeElapsed(Instant created) {
         return created.plus(sessionLifetime).isBefore(clock.instant());
     }
 
+    public void deleteExpiredRemoteAndLocalSessions(Clock clock, Predicate<Session> sessionIsActiveForApplication) {
+        // All known sessions, both local (file) and remote (zookeeper)
+        Set<Long> sessions = getLocalSessionsIdsFromFileSystem();
+        sessions.addAll(getRemoteSessionsFromZooKeeper());
+
+        log.log(Level.FINE, () -> "Sessions for tenant " + tenantName + ": " + sessions);
+
+        // TODO: Sessions created last 30 seconds are considered new, exclude these to avoid deleting new sessions
+        Set<Long> newSessions = findNewSessionsInFileSystem();
+        // Avoid deleting too many in one run
+        int deleteMax = (int) Math.min(1000, Math.max(50, sessions.size() * 0.05));
+        int deleted = 0;
+        try {
+            for (Long sessionId : sessions) {
+                // Skip sessions newly added (we might have a session in the file system, but not in ZooKeeper,
+                // we will exclude these)
+                if (newSessions.contains(sessionId)) continue;
+
+                Session session = remoteSessionCache.get(sessionId);
+                if (session == null)
+                    session = new RemoteSession(tenantName, sessionId, createSessionZooKeeperClient(sessionId));
+
+                Instant createTime = session.getCreateTime();
+                Session.Status status = session.getStatus();
+                if (session.getStatus() == ACTIVATE && sessionIsActiveForApplication.test(session)) continue;
+
+                boolean hasExpired = hasExpired(createTime);
+                if (hasExpired) {
+                    log.log(Level.FINE, () -> "Remote session " + sessionId + " for " + tenantName + " has expired, deleting it");
+                    deleteRemoteSessionFromZooKeeper(session);
+                    deleted++;
+                }
+
+                log.log(Level.FINE, () -> "Candidate local session for deletion: " + sessionId +
+                        ", created: " + createTime + ", status " + status + ", can be deleted: " + canBeDeleted(sessionId, status) +
+                        ", hasExpired: " + hasExpired);
+                if (hasExpired && canBeDeleted(sessionId, status)) {
+                    log.log(Level.FINE, () -> "expired: " + hasExpired + ", can be deleted: " + canBeDeleted(sessionId, status));
+                    deleteLocalSession(sessionId);
+                } else if (createTime.plus(Duration.ofDays(1)).isBefore(clock.instant())) {
```

Comment:
Could expiration time ever exceed 1 day? If it does, this becomes a bit weird—it's not "expired", but we attempt forceful deletion anyway.